kubetest2/pkg/build/build.go

```diff
@@ -0,0 +1,147 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package build implements a common system for building kubernetes for deployers to use.
+package build
+
+import (
+	"fmt"
+	"go/build"
+	"io/ioutil"
+	"log"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"regexp"
+	"strings"
+)
+
+const (
+	target = "bazel-release"
+)
+
+var (
+	// This will need changed to support other platforms.
+	tarballsToExtract = []string{
+		"kubernetes.tar.gz",
+		"kubernetes-test-linux-amd64.tar.gz",
+		"kubernetes-test-portable.tar.gz",
+		"kubernetes-client-linux-amd64.tar.gz",
+	}
+)
+
+// Build builds kubernetes with the bazel-release target
+func Build() error {
+	// TODO(RonWeber): This needs options
+	goPath := os.Getenv("GOPATH")
+	k8ssrc := filepath.Join(goPath, "src", "k8s.io", "kubernetes")
+	c := exec.Command("make", "-C", k8ssrc, target)
+	inheritOutput(c)
+	err := c.Run()
+	if err != nil {
+		return err
+	}
+	return extractBuiltTars()
+}
+
+// Stage stages the build to GCS.
+// Essentially release/push-build.sh --bucket=B --ci --gcs-suffix=S --noupdatelatest
+func Stage(location string) error {
+	re := regexp.MustCompile(`^gs://([\w-]+)/(devel|ci)(/.*)?`)
+	mat := re.FindStringSubmatch(location)
+	if mat == nil {
+		return fmt.Errorf("Invalid stage location: %v. Use gs://bucket/ci/optional-suffix", location)
+	}
+	bucket := mat[1]
+	ci := mat[2] == "ci"
+	gcsSuffix := mat[3]
+
+	args := []string{
+		"--nomock",
+		"--verbose",
+		"--noupdatelatest",
+		fmt.Sprintf("--bucket=%v", bucket),
+	}
+	if len(gcsSuffix) > 0 {
+		args = append(args, fmt.Sprintf("--gcs-suffix=%v", gcsSuffix))
+	}
+	if ci {
+		args = append(args, "--ci")
+	}
+
+	name, err := K8sDir("release", "push-build.sh")
+	if err != nil {
+		return err
+	}
+	cmd := exec.Command(name, args...) //Not using kubetest2/pkg/exec because we need to set cmd.Dir.
+	cmd.Dir, err = K8sDir("kubernetes")
+	inheritOutput(cmd)
+	if err != nil {
+		return err
+	}
+	return cmd.Run()
+
+}
+
+// K8sDir returns $GOPATH/src/k8s.io/...
+func K8sDir(topdir string, parts ...string) (string, error) {
+	gopathList := filepath.SplitList(build.Default.GOPATH)
+	for _, gopath := range gopathList {
+		kubedir := filepath.Join(gopath, "src", "k8s.io", topdir)
+		if _, err := os.Stat(kubedir); !os.IsNotExist(err) {
+			p := []string{kubedir}
+			p = append(p, parts...)
+			return filepath.Join(p...), nil
+		}
+	}
+	return "", fmt.Errorf("could not find directory src/k8s.io/%s in GOPATH", topdir)
+}
+
+// TODO(RonWeber): This whole untarring and cd-ing logic is out of
+// scope for Build().  It needs a better home.
+func extractBuiltTars() error {
+	allBuilds, err := K8sDir("kubernetes", "_output", "gcs-stage")
+	if err != nil {
+		return err
+	}
+	ls, err := ioutil.ReadDir(allBuilds)
+	if err != nil {
+		return err
+	}
+	for _, dir := range ls {
+		dirName := dir.Name()
+		if strings.HasPrefix(dirName, "v") { //This is probably the release we just built.
```

Comment:
/shrug - seems fragile..

Reply:
This is the logic kubetest uses now (I think).  If you have a better idea, I'm open to suggestions.

Reply:
less fragile would probably be identifying a valid version file

Reply:
Then I would have to peek into the tarball, and if I'm doing that I might as well just extract the tarball and do a sanity check afterwards.

Reply:
/this-is-fine

Reply:
The only real thing this check does is make sure we don't extract the same tar multiple times.  I'm going to write another version that makes that more explicit.  It's still really fragile.